### PR TITLE
add `extend_from_slice` method

### PR DESCRIPTION
**Pull request by vxpm**
adds an `extend_from_slice` method to SegVec (and fixes a small bug in `extend` - it was reserving elements _twice_).

needed for acceptable performance in #40

### CODE REVIEW

File: src/tests.rs
Comment (mccolljr) [inline src/tests.rs]: I think we probably want to validate all the elements of `sv` here. My reading of this code is that we're just validating the first `N` items, where `N` is the length of `elements`. You could do something like:

```suggestion
    for (i, v) in sv.iter().enumerate() {
        assert_eq!(v, &elements[i%elements.len()]);
    }
```

File: src/lib.rs
Comment (mccolljr) [inline src/lib.rs]: Unfortunately, I don't think we can do this.

The docs for [`size_hint`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint) say:

> size_hint() is primarily intended to be used for optimizations such as reserving space for the elements of the iterator, but must not be trusted to e.g., omit bounds checks in unsafe code. An incorrect implementation of size_hint() should not lead to memory safety violations.

So - this code can call `self.reserve(additional)` with the result of caling `size_hint` as an optimization, but we can't treat this as a guaratee that `iter` will only produce exactly `additional` elements.

**Review by mccolljr — CHANGES_REQUESTED** [review]
I've been extremely busy with work and have let these PRs linger. Sorry for the crazy delay on this.

This change looks mostly great, but there's one safety issue that we will need to correct before I can hit the merge button. I promise to be more responsive to the next round of changes, I've got time off at the end of the year that I'll be using to catch up on personal projects like this.

File: src/lib.rs
Comment (mccolljr) [reply src/lib.rs]: We could do one of two things:

1. Revert to the `push` implementation that was there previously, or
2. Use `push_unchecked` until we've exhausted the reserved space, then fall back to `push`

I assume the most common case will be that `size_hint` is correct and honest, and I assume that `push_count < additional` and `push_count += 1` are still faster than a checked `push`, so it seems like option 2 might be worth it. I haven't had a chance to actually benchmark this to make sure, though.

File: src/tests.rs
Comment (vxpm) [reply src/tests.rs]: oops, you're right! totally slipped past me

File: src/lib.rs
Comment (vxpm) [reply src/lib.rs]: that's true - somehow i forgot that `size_hint` is, indeed, just a hint. i think option 2 is the most reasonable, so i'll implement it and see how it goes.

**Comment by vxpm** [comment]
it's alright! nothing urgent. i'm currently not using this crate actively on any projects - i just think this is a good, useful data structure and want to help it get faster & closer to feature parity with good old `Vec`. i might be using it soon, though, so do expect some more PRs in this end of year.

**Comment by vxpm** [comment]
i believe it's all good now :)
edit: by the way, fixing the unsoundness in `extend` seems to have impacted performance positively - not sure why, but that's very welcome.

**Comment by mccolljr** [comment]
Awesome, thanks for the changes. Merged!

**Pull request merged**
